File: app/src/test/java/linusfessler/alarmtiles/TimeFormatterTest.java
package linusfessler.alarmtiles;

import org.junit.Assert;
import org.junit.Test;

import java.util.concurrent.TimeUnit;

public class TimeFormatterTest {

    private final TimeFormatter timeFormatter = new TimeFormatter();

    @Test
    public void formatMillisWithPrecisionMilliseconds() {
        // GIVEN
        final int millis = 123456789;
        final TimeUnit precision = TimeUnit.MILLISECONDS;
        final String expected = "34h 17m 36s 78";
        // WHEN
        final String actual = this.timeFormatter.format(millis, precision);
        // THEN
        Assert.assertEquals(expected, actual);
    }

    @Test
    public void formatMillisWithPrecisionSeconds() {
        // GIVEN
        final int millis = 123456789;
        final TimeUnit precision = TimeUnit.SECONDS;
        final String expected = "34h 17m 36s";
        // WHEN
        final String actual = this.timeFormatter.format(millis, precision);
        // THEN
        Assert.assertEquals(expected, actual);
    }

    @Test
    public void formatMillisWithPrecisionMinutes() {
        // GIVEN
        final int millis = 123456789;
        final TimeUnit precision = TimeUnit.MINUTES;
        final String expected = "34h 17m";
        // WHEN
        final String actual = this.timeFormatter.format(millis, precision);
        // THEN
        Assert.assertEquals(expected, actual);
    }

    @Test
    public void formatMillisWithPrecisionHours() {
        // GIVEN
        final int millis = 123456789;
        final TimeUnit precision = TimeUnit.HOURS;
        final String expected = "34h";
        // WHEN
        final String actual = this.timeFormatter.format(millis, precision);
        // THEN
        Assert.assertEquals(expected, actual);
    }

    @Test
    public void formatHoursMinutesSecondsMillis() {
        // GIVEN
        final int hours = 4;
        final int minutes = 15;
        final int seconds = 30;
        final int millis = 450;
        final String expected = "4h 15m 30s 45";
        // WHEN
        final String actual = this.timeFormatter.format(hours, minutes, seconds, millis);
        // THEN
        Assert.assertEquals(expected, actual);
    }

    @Test
    public void formatHoursMinutesSeconds() {
        // GIVEN
        final int hours = 4;
        final int minutes = 15;
        final int seconds = 30;
        final String expected = "4h 15m 30s";
        // WHEN
        final String actual = this.timeFormatter.format(hours, minutes, seconds);
        // THEN
        Assert.assertEquals(expected, actual);
    }

    @Test
    public void formatHoursMinutes() {
        // GIVEN
        final int hours = 4;
        final int minutes = 15;
        final String expected = "4h 15m";
        // WHEN
        final String actual = this.timeFormatter.format(hours, minutes);
        // THEN
        Assert.assertEquals(expected, actual);
    }

    @Test
    public void formatHours() {
        // GIVEN
        final int hours = 4;
        final int minutes = 0;
        final int seconds = 0;
        final int millis = 0;
        final String expected = "4h 0m 0s 00";
        // WHEN
        final String actual = this.timeFormatter.format(hours, minutes, seconds, millis);
        // THEN
        Assert.assertEquals(expected, actual);
    }

    @Test
    public void formatMinutes() {
        // GIVEN
        final int hours = 0;
        final int minutes = 15;
        final int seconds = 0;
        final int millis = 0;
        final String expected = "15m 0s 00";
        // WHEN
        final String actual = this.timeFormatter.format(hours, minutes, seconds, millis);
        // THEN
        Assert.assertEquals(expected, actual);
    }

    @Test
    public void formatSeconds() {
        // GIVEN
        final int hours = 0;
        final int minutes = 0;
        final int seconds = 30;
        final int millis = 0;
        final String expected = "30s 00";
        // WHEN
        final String actual = this.timeFormatter.format(hours, minutes, seconds, millis);
        // THEN
        Assert.assertEquals(expected, actual);
    }

    @Test
    public void formatMillis() {
        // GIVEN
        final int hours = 0;
        final int minutes = 0;
        final int seconds = 0;
        final int millis = 450;
        final String expected = "45";
        // WHEN
        final String actual = this.timeFormatter.format(hours, minutes, seconds, millis);
        // THEN
        Assert.assertEquals(expected, actual);
    }
}

File: app/src/test/java/linusfessler/alarmtiles/TimeFormatterTest.kt
package linusfessler.alarmtiles

import linusfessler.alarmtiles.shared.TimeFormatter
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test
import java.util.concurrent.TimeUnit

class TimeFormatterTest {
    private val timeFormatter = TimeFormatter()

    @Test
    fun formatMillisWithPrecisionMilliseconds() {
        // GIVEN
        val millis = 123456789
        val precision = TimeUnit.MILLISECONDS
        val expected = "34h 17m 36s 78"
        // WHEN
        val actual = timeFormatter.format(millis.toLong(), precision)
        // THEN
        Assertions.assertEquals(expected, actual)
    }

    @Test
    fun formatMillisWithPrecisionSeconds() {
        // GIVEN
        val millis = 123456789
        val precision = TimeUnit.SECONDS
        val expected = "34h 17m 36s"
        // WHEN
        val actual = timeFormatter.format(millis.toLong(), precision)
        // THEN
        Assertions.assertEquals(expected, actual)
    }

    @Test
    fun formatMillisWithPrecisionMinutes() {
        // GIVEN
        val millis = 123456789
        val precision = TimeUnit.MINUTES
        val expected = "34h 17m"
        // WHEN
        val actual = timeFormatter.format(millis.toLong(), precision)
        // THEN
        Assertions.assertEquals(expected, actual)
    }

    @Test
    fun formatMillisWithPrecisionHours() {
        // GIVEN
        val millis = 123456789
        val precision = TimeUnit.HOURS
        val expected = "34h"
        // WHEN
        val actual = timeFormatter.format(millis.toLong(), precision)
        // THEN
        Assertions.assertEquals(expected, actual)
    }

    @Test
    fun formatHoursMinutesSecondsMillis() {
        // GIVEN
        val hours = 4
        val minutes = 15
        val seconds = 30
        val millis = 450
        val expected = "4h 15m 30s 45"
        // WHEN
        val actual = timeFormatter.format(hours.toLong(), minutes.toLong(), seconds.toLong(), millis.toLong())
        // THEN
        Assertions.assertEquals(expected, actual)
    }

    @Test
    fun formatHoursMinutesSeconds() {
        // GIVEN
        val hours = 4
        val minutes = 15
        val seconds = 30
        val expected = "4h 15m 30s"
        // WHEN
        val actual = timeFormatter.format(hours.toLong(), minutes.toLong(), seconds.toLong())
        // THEN
        Assertions.assertEquals(expected, actual)
    }

    @Test
    fun formatHoursMinutes() {
        // GIVEN
        val hours = 4
        val minutes = 15
        val expected = "4h 15m"
        // WHEN
        val actual = timeFormatter.format(hours.toLong(), minutes.toLong())
        // THEN
        Assertions.assertEquals(expected, actual)
    }

    @Test
    fun formatHours() {
        // GIVEN
        val hours = 4
        val minutes = 0
        val seconds = 0
        val millis = 0
        val expected = "4h 0m 0s 00"
        // WHEN
        val actual = timeFormatter.format(hours.toLong(), minutes.toLong(), seconds.toLong(), millis.toLong())
        // THEN
        Assertions.assertEquals(expected, actual)
    }

    @Test
    fun formatMinutes() {
        // GIVEN
        val hours = 0
        val minutes = 15
        val seconds = 0
        val millis = 0
        val expected = "15m 0s 00"
        // WHEN
        val actual = timeFormatter.format(hours.toLong(), minutes.toLong(), seconds.toLong(), millis.toLong())
        // THEN
        Assertions.assertEquals(expected, actual)
    }

    @Test
    fun formatSeconds() {
        // GIVEN
        val hours = 0
        val minutes = 0
        val seconds = 30
        val millis = 0
        val expected = "30s 00"
        // WHEN
        val actual = timeFormatter.format(hours.toLong(), minutes.toLong(), seconds.toLong(), millis.toLong())
        // THEN
        Assertions.assertEquals(expected, actual)
    }

    @Test
    fun formatMillis() {
        // GIVEN
        val hours = 0
        val minutes = 0
        val seconds = 0
        val millis = 450
        val expected = "45"
        // WHEN
        val actual = timeFormatter.format(hours.toLong(), minutes.toLong(), seconds.toLong(), millis.toLong())
        // THEN
        Assertions.assertEquals(expected, actual)
    }

    @Test
    fun formatZeroMillis() {
        // GIVEN
        val hours = 0
        val minutes = 0
        val seconds = 0
        val millis = 0
        val expected = "00"
        // WHEN
        val actual = timeFormatter.format(hours.toLong(), minutes.toLong(), seconds.toLong(), millis.toLong())
        // THEN
        Assertions.assertEquals(expected, actual)
    }

    @Test
    fun formatZeroSeconds() {
        // GIVEN
        val hours = 0
        val minutes = 0
        val seconds = 0
        val expected = "0s"
        // WHEN
        val actual = timeFormatter.format(hours.toLong(), minutes.toLong(), seconds.toLong())
        // THEN
        Assertions.assertEquals(expected, actual)
    }

    @Test
    fun formatZeroMinutes() {
        // GIVEN
        val hours = 0
        val minutes = 0
        val expected = "0m"
        // WHEN
        val actual = timeFormatter.format(hours.toLong(), minutes.toLong())
        // THEN
        Assertions.assertEquals(expected, actual)
    }

    @Test
    fun formatZeroHours() {
        // GIVEN
        val hours = 0
        val expected = "0h"
        // WHEN
        val actual = timeFormatter.formatHours(hours.toLong())
        // THEN
        Assertions.assertEquals(expected, actual)
    }
}
